src/fparser/sourceinfo.py

```diff
@@ -62,114 +62,147 @@
-
-_has_f_extension = re.compile(r'.*[.](for|ftn|f77|f)\Z',re.I).match
-_has_f_header = re.compile(r'-[*]-\s*(fortran|f77)\s*-[*]-',re.I).search
-_has_f90_header = re.compile(r'-[*]-\s*f90\s*-[*]-',re.I).search
-_has_f03_header = re.compile(r'-[*]-\s*f03\s*-[*]-',re.I).search
-_has_f08_header = re.compile(r'-[*]-\s*f08\s*-[*]-',re.I).search
-_has_free_header = re.compile(r'-[*]-\s*(f90|f95|f03|f08)\s*-[*]-',re.I).search
-_has_fix_header = re.compile(r'-[*]-\s*fix\s*-[*]-',re.I).search
-_has_pyf_header = re.compile(r'-[*]-\s*pyf\s*-[*]-',re.I).search
-_free_format_start = re.compile(r'[^c*!]\s*[^\s\d\t]',re.I).match
-
-def get_source_info(filename):
-    """
-    Determine if fortran file is
-      - in fix format and contains Fortran 77 code    -> return False, True
-      - in fix format and contains Fortran 90,95,2003,2008 code    -> return False, False
-      - in free format and contains Fortran 90,95,2003,2008 code   -> return True, False
-      - in free format and contains signatures (.pyf) -> return True, True
-    """
-    base,ext = os.path.splitext(filename)
-    if ext=='.pyf':
-        return True, True
-    isfree = False
-    isstrict = False
-    f = open(filename,'r')
-    firstline = f.readline()
-    f.close()
-    if _has_f_header(firstline): return False, True
-    if _has_fix_header(firstline): return False, False
-    if _has_free_header(firstline): return True, False
-    if _has_pyf_header(firstline): return True, True
-    if _has_f_extension(filename) and \
-       not (_has_free_header(firstline) or _has_fix_header(firstline)):
-        isstrict = True
-    elif is_free_format(filename) and not _has_fix_header(firstline):
-        isfree = True
-    return isfree,isstrict
-
-def get_source_info_str(code):
-    """
-    Determine the format of Fortran code.
-    See get_source_info() for the meaning of return values.
-    """
-    firstline = code.lstrip().split('\n',1)[0]
-    if _has_f_header(firstline): return False, True
-    if _has_fix_header(firstline): return False, False
-    if _has_free_header(firstline): return True, False
-    if _has_pyf_header(firstline): return True, True
-    n = 10000 # the number of non-comment lines to scan for hints
-    lines = code.splitlines()
-    isfree = False
-    while n>0 and lines:
+
+
+##############################################################################
+
+class FortranFormat(object):
+    '''
+    Describes the nature of a piece of Fortran source.
+
+    Source can be fixed or free format. It can also be "strict" or
+    "not strict" although it's not entirely clear what that means.
+    '''
+    def __init__(self, is_free, is_strict):
+        self._is_free = is_free
+        self._is_strict = is_strict
+
+    def __eq__(self, other):
+        if isinstance(other, FortranFormat):
+            return self._is_free == other.is_free() \
+                   and self._is_strict == other.is_strict()
+        raise NotImplementedError
+
+    def __repr__(self):
```

Comment:
The body of this routine is not covered by the test suite.

Reply:
While writing tests for it I decided that it should be a `__str__`, not a `__repr__`, following the rule of thumb: Repr to uniquely identify, Str for users to read.